include/lbann/execution_contexts/execution_context.hpp

```diff
@@ -0,0 +1,151 @@
+  }
+
+  /** Archive for checkpoint and restart */
+  template <class Archive> void serialize( Archive & ar ) {
+    ar(CEREAL_NVP(m_execution_mode),
+       CEREAL_NVP(m_terminate_training),
+       CEREAL_NVP(m_step));
+  }
+
+  /** @brief Current step in the training algorithm
+    *  @detailed Step counts the number of iterations in the training
+    *  algorithm's internal state
+    */
+  El::Int get_step() const noexcept { return m_step; }
+
+  /** @brief Increment the current step in the training algorithm
+    *  @detailed Increment the step count in the training
+    *  algorithm's internal state
+    */
+  void inc_step() noexcept { ++m_step; }
+
+  /** Get the mode that the trainer is currenting executing. */
+  inline void set_execution_mode(execution_mode mode) noexcept {
+    m_execution_mode = mode;
+  }
+
+  /** Get the mode that the trainer is currenting executing. */
+  inline execution_mode get_execution_mode() const noexcept {
+    return m_execution_mode;
+  }
+
+  /** Return true if the flag to stop training is set. */
+  bool get_terminate_training() const {
+    return m_terminate_training;
+  }
+  /** Set the terminate training flag (on or off). */
+  void set_terminate_training(bool f) {
+    m_terminate_training = f;
+  }
+
+  /** Get the execution environment */
+  trainer* get_trainer() {
+    return m_trainer;
+  }
+
+  observing_ptr<thread_pool> get_io_thread_pool() const;
+
+  observing_ptr<lbann_comm> get_comm() const { return m_comm; };
+
+  /** Are background I/O activities enabled by the input layers */
+  bool background_io_activity_allowed();
+
+  /** Checkpoint training_algorithm to given file descriptor, return number of bytes written */
+  virtual bool save_to_checkpoint_shared(persist& p);
+  /** Restore training_algorithm by reading checkpoint from given file descriptor, return number of bytes read */
+  virtual bool load_from_checkpoint_shared(persist& p);
+  virtual bool save_to_checkpoint_distributed(persist& p);
+  virtual bool load_from_checkpoint_distributed(persist& p);
+
+protected:
+  /** Copy constructor. */
+  execution_context(const execution_context& other) = default;
+  /** Copy assignment operator. */
+  execution_context& operator=(const execution_context& other) = default;
+  /** Move constructor. */
+  execution_context(execution_context&& other) = default;
+  /** Move assignment operator. */
+  execution_context& operator=(execution_context&& other) = default;
+
+private:
+  /** Pointer to the training context (execution environment) for the training algorithm */
+  observing_ptr<trainer> m_trainer;
+
+  /** LBANN communicator. */
+  observing_ptr<lbann_comm> m_comm;
+
+  /** The trainer's current execution mode. */
+  execution_mode m_execution_mode = execution_mode::training;
+
+  /** @brief Current step in the training algorithm
+    *  @detailed Step counts the number of iterations in the training
+    *  algorithm's internal state
+    */
+  El::Int m_step = 0;
```

Comment:
IMO, this should be a `size_t`. The reasons are these: one can not have a negative number of steps; one might do a lot of steps; and this does not directly interact with `El::` objects so there's no reason to tie its type to that library.